Add `HashTable` methods related to the raw bucket index

cuviper opened the pull request:
On `HashTable<T, A>`:
```rust
pub fn num_buckets(&self) -> usize;
pub fn find_bucket_index(&self, hash: u64, eq: impl FnMut(&T) -> bool) -> Option<usize>;
pub fn get_bucket_entry(&mut self, index: usize) -> Result<OccupiedEntry<'_, T, A>, AbsentEntry<'_, T, A>>;
pub unsafe fn get_bucket_entry_unchecked(&mut self, index: usize) -> OccupiedEntry<'_, T, A>;
pub fn get_bucket(&self, index: usize) -> Option<&T>;
pub unsafe fn get_bucket_unchecked(&self, index: usize) -> &T;
pub fn get_bucket_mut(&mut self, index: usize) -> Option<&mut T>;
pub unsafe fn get_bucket_unchecked_mut(&mut self, index: usize) -> &mut T;
pub fn iter_buckets(&self) -> IterBuckets<'_, T>;
pub fn iter_hash_buckets(&self, hash: u64) -> IterHashBuckets<'_, T>;
```

On `OccupiedEntry<'_, T, A>`:
```rust
pub fn bucket_index(&self) -> usize;
```

`IterHashBuckets<'_, T>` implements `Clone`, `Debug`, `Default`, `Iterator<Item = usize>`, and `FusedIterator`.
`IterBuckets<'_, T>` implements all those and also `ExactSizeIterator`.

Closes #613, although I chose different names.

clarfonthey commented on src/table.rs:
````diff
@@ -591,6 +755,44 @@ where
+    /// #     test()
+    /// # }
+    /// ```
+    pub fn buckets(&self) -> usize {
````
I personally would name this `num_buckets` rather than `buckets` since the name as-is makes me think that it's returning an iterator instead.

clarfonthey reviewed (changes requested):
I think that `get_bucket` and `get_bucket_mut` are good, but `buckets` should be renamed to `num_buckets` to make it clear it's returning a count and not an iterator.

In terms of `find_bucket_index` and `get_bucket_entry`, I left some comments. But if you wanted to merge the smaller set first, I'm fine with that.

clarfonthey commented:
re: unsafe methods, I guess the main concern is whether the bounds checks can easily be optimised out. While there isn't the same robust codegen testing as the compiler here, maybe it would be worth poking around on godbolt and seeing how `find_bucket_index` followed by one of the relevant methods is handled.

cuviper commented:
I assume you mean adding that in `find_bucket_index`?

```rust
    pub fn find_bucket_index(&self, hash: u64, eq: impl FnMut(&T) -> bool) -> Option<usize> {
        match self.raw.find(hash, eq) {
            Some(bucket) => Some(unsafe {
                let index = self.raw.bucket_index(&bucket);
                core::hint::assert_unchecked(index < self.raw.buckets());
                core::hint::assert_unchecked(self.raw.is_bucket_full(index));
                index
            }),
            None => None,
        }
    }
```

It does remove the unwrap in my simple back-to-back test, but it also pessimizes the unchecked case a little, recomputing the final address. Different variations of that assertion had the same effect.

```asm
+        neg r9
+        lea rax, [rax + 8*r9]
         add rax, -8
         pop rcx
         .cfi_def_cfa_offset 8
         ret
```

In a real scenario I expect you'd also have other interim code, and who knows if the assertion could still be propagated. I don't think we should bother until there's a real measured benefit, and anyway even the checked case is still much less than a full hash probe that you would have needed before!

cuviper commented:
RE MSRV, I recently bumped `indexmap` to 1.82, which is just over a year old.

cuviper commented on src/raw/mod.rs:
```diff
@@ -11,6 +11,8 @@ use core::ptr::NonNull;
 use core::slice;
 use core::{hint, ptr};
 
+pub(crate) use crate::control::Tag;
```
Because `table::VacantEntry` now has a field of this type.

clarfonthey replied:
Yeah, but why re-export instead of just importing in that file too? I prefer to avoid having too many aliases because it makes the code difficult to follow.

cuviper replied:
Oh, right. I was thinking it came from a submodule here, and I clearly didn't pay attention to the path. I've now made `table` use that directly, thanks!

cuviper commented:
> Is there any downside to having `get_bucket_mut` always return an `OccupiedEntry`? It's strictly more flexible than returning a `&mut T` and this API is only intended for advanced use cases anyways. Same with the unchecked version.

i.e. instead of the separate `get_bucket_entry`, I suppose? At first I needed the hash value for that, which was a downside if you only wanted `&mut T`, but now I guess they are functionally equivalent.

I suppose the downside now is just that the caller would have to use an extra `.get_mut()` or `.into_mut()` on the entry if that's all they want, and hope that the compiler does optimize the unused stuff away. It probably will though.

Hmm, if `OccupiedEntry` implemented `Deref<Target = T>` and `DerefMut`, then my experimental branch in `indexmap` wouldn't even need to change the way it's currently using `get_bucket_mut()`. That could be nice, but it's not required.

cuviper commented:
Note that there is a similar trio in `find`, `find_mut`, and `find_entry` as well.

cuviper commented:
I went ahead and made that change to try it -- easy enough to drop that commit if we change our minds. :)

cuviper commented:
* Reverted `get_bucket_mut` (and unchecked) back to `&mut T`
* Changed `get_bucket_entry` to return `Result<OccupiedEntry, AbsentEntry>`
* Added `get_bucket_entry_unchecked` returning `OccupiedEntry`
* Added `iter_buckets` and `iter_hash_buckets`

> The only missing functionality here is the ability to insert a value just by bucket index, but it's not clear how useful this would be since the only way to get a valid index for this is by re-using the index of a previous insertion of the same key.

I don't think we should support this, as it would let you insert values in places that wouldn't be probed by the same hash, especially if we're not making any guarantees about index stability after removals. You *can* do something like:

```rust
    let (old, vacant) = table.get_bucket_entry(index).unwrap().remove();
    let new_index = vacant.insert(new).bucket_index();
```

... but I think even there we're not _promising_ that the new index will be the same. Supposing we were to shift items to avoid tombstones, then the vacant entry might already represent a different bucket.

cuviper commented on src/table.rs:
```diff
@@ -741,6 +1071,42 @@ where
+    /// An iterator producing the `usize` indices of all occupied buckets.
+    ///
+    /// The order in which the iterator yields indices is unspecified
+    /// and may change in the future.
```
I think it would actually be fine to guarantee increasing order here, but the underlying `raw::FullBucketsIndices` is explicitly disclaiming that.

Amanieu replied:
I think it's fine to leave it unspecified for now unless there's a use case that clearly asks for it.

cuviper commented:
Rebased to resolve conflicts with the `InsertSlot` removal.

Amanieu commented on src/table.rs:
```diff
@@ -2230,6 +2727,32 @@ where
+impl Iterator for IterHashBuckets<'_> {
+    type Item = usize;
+
+    fn next(&mut self) -> Option<Self::Item> {
```
This method is not generic and need an `#[inline]`.
